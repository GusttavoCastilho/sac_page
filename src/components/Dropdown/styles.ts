import styled, { css } from "styled-components";
import { lighten } from "polished";
import media from "styled-media-query";

type WrapperProps = {
  isOpen?: boolean;
};

const WrapperModifers = {
  open: () => css`
    opacity: 1;
    height: 100px;
    pointer-events: auto;
    transform: translateY(0);

    ${media.lessThan("small")`
      height: 60px;
    `}
  `,
  close: () => css`
    opacity: 0;
    pointer-events: none;
    transform: translateY(-2rem);
  `,
};

export const Wrapper = styled.div<WrapperProps>`
  ${({ isOpen, theme }) => css`
    position: relative;
    width: 80%;

    ${Content} {
      transition: transform 0.2s ease-in, opacity 0.3s ease-in-out;
      ${isOpen && WrapperModifers.open()}
      ${!isOpen && WrapperModifers.close()}
    }

    ${media.lessThan("large")`
     width: 100%;
     padding: 0 ${theme.spacings.xxsmall};
     height: ${isOpen ? "100px" : "50px"};
    `}
  `}
`;

export const Title = styled.h5`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.large};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.purple};

    display: flex;
    align-items: center;

    cursor: pointer;
    position: relative;
    padding-right: 2.4rem;

    ${media.lessThan("large")`
      font-size: ${theme.font.sizes.medium};
      padding-right: 10px;
    `}

    ${media.lessThan("medium")`
      font-size: ${theme.font.sizes.small};
      padding-right: 10px;
    `}
  `}
`;

export const Content = styled.p`
  ${({ theme }) => css`
    width: 100%;

    margin-top: ${theme.spacings.small};

    font-size: ${theme.font.sizes.medium};
    color: ${theme.colors.dark};

    line-height: ${theme.spacings.small};

    ${media.lessThan("large")`
      font-size: ${theme.font.sizes.small};
      margin-top: ${theme.spacings.xxsmall};
      line-height: ${theme.spacings.xsmall};
    `}

    ${media.lessThan("medium")`
      font-size: ${theme.font.sizes.xsmall};
      margin-top: ${theme.spacings.xxsmall};
      line-height: ${theme.spacings.xsmall};
    `}
  `}
`;

export const BoxTitle = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;

    div {
      flex: 1;
      background-color: ${lighten(0.4, theme.colors.dark)};
      height: 1px;
    }
  `}
`;
