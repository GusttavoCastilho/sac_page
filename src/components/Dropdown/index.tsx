import { useState } from "react";
import { useTheme } from "styled-components";

import {
  ArrowIosDownwardOutline,
  ArrowIosUpwardOutline,
} from "@styled-icons/evaicons-outline";
import * as S from "./styles";

export type DropdownProps = {
  title: string;
  children: string;
};

const Dropdown = ({ title, children }: DropdownProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const theme = useTheme();
  return (
    <S.Wrapper isOpen={isOpen}>
      <S.BoxTitle>
        <S.Title onClick={() => setIsOpen(!isOpen)}>{title}</S.Title>
        <div />
        {isOpen ? (
          <ArrowIosUpwardOutline
            color={theme.colors.purple}
            size={28}
            onClick={() => setIsOpen(!isOpen)}
            style={{ cursor: "pointer" }}
            aria-label="arrow-up"
            aria-hidden={!isOpen}
          />
        ) : (
          <ArrowIosDownwardOutline
            color={theme.colors.purple}
            size={28}
            onClick={() => setIsOpen(!isOpen)}
            style={{ cursor: "pointer" }}
            aria-label="arrow-down"
            aria-hidden={!isOpen}
          />
        )}
      </S.BoxTitle>
      <S.Content aria-hidden={!isOpen}>{children}</S.Content>
    </S.Wrapper>
  );
};
export default Dropdown;
