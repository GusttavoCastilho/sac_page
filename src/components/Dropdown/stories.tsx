import React from "react";
import { Story, Meta } from "@storybook/react/types-6-0";
import Dropdown, { DropdownProps } from ".";

export default {
  title: "Dropdown",
  component: Dropdown,
  argTypes: {
    title: {
      type: "string",
    },
    children: {
      type: "string",
    },
  },
} as Meta;

export const Default: Story<DropdownProps> = () => (
  <Dropdown title="O catálogo é grátis?">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed a ligula risus.
    Quisque sagittis nunc ut diam sagittis tempor. Etiam mollis, felis a
    consectetur volutpat,
  </Dropdown>
);
