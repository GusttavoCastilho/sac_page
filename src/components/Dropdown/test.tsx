import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme } from "../../utils/helpers";
import Dropdown from ".";

describe("<Button />", () => {
  it("should render title and children", () => {
    const { container } = renderWithTheme(
      <Dropdown title="my title">my children</Dropdown>
    );

    expect(
      screen.getByRole("heading", { name: /my title/i })
    ).toBeInTheDocument();
    expect(screen.getByText(/my children/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should handle open/close dropdown", () => {
    renderWithTheme(<Dropdown title="my title">my children</Dropdown>);

    const description = screen.getByText(/my children/i);

    expect(description).toHaveStyle({ opacity: 0 });
    expect(description.getAttribute("aria-hidden")).toBe("true");

    userEvent.click(screen.getByText(/my title/i));

    expect(description).toHaveStyle({ opacity: 1 });
    expect(description.getAttribute("aria-hidden")).toBe("false");
  });

  it("should handle icon from open/close dropdown", () => {
    renderWithTheme(<Dropdown title="my title">my children</Dropdown>);

    expect(
      screen.getByLabelText(/arrow-down/i).getAttribute("aria-hidden")
    ).toBe("true");

    userEvent.click(screen.getByLabelText(/arrow-down/i));

    expect(screen.getByLabelText(/arrow-up/i).getAttribute("aria-hidden")).toBe(
      "false"
    );
  });
});
