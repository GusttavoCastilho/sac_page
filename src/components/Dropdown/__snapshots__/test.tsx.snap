// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should render title and children 1`] = `
.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  width: 80%;
}

.c0 .c4 {
  -webkit-transition: -webkit-transform 0.2s ease-in,opacity 0.3s ease-in-out;
  -webkit-transition: transform 0.2s ease-in,opacity 0.3s ease-in-out;
  transition: transform 0.2s ease-in,opacity 0.3s ease-in-out;
  opacity: 0;
  pointer-events: none;
  -webkit-transform: translateY(-2rem);
  -ms-transform: translateY(-2rem);
  transform: translateY(-2rem);
}

.c2 {
  font-size: 2.0rem;
  font-weight: 600;
  color: #A141EB;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  padding-right: 2.4rem;
}

.c5 {
  width: 100%;
  margin-top: 2.4rem;
  font-size: 1.6rem;
  color: #5B5B5B;
  line-height: 2.4rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 div {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #c1c1c1;
  height: 1px;
}

@media (max-width:1170px) {
  .c0 {
    width: 100%;
    padding: 0 0.8rem;
    height: 50px;
  }
}

@media (max-width:1170px) {
  .c2 {
    font-size: 1.6rem;
    padding-right: 10px;
  }
}

@media (max-width:768px) {
  .c2 {
    font-size: 1.4rem;
    padding-right: 10px;
  }
}

@media (max-width:1170px) {
  .c5 {
    font-size: 1.4rem;
    margin-top: 0.8rem;
    line-height: 1.6rem;
  }
}

@media (max-width:768px) {
  .c5 {
    font-size: 1.2rem;
    margin-top: 0.8rem;
    line-height: 1.6rem;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <h5
      class="c2"
    >
      my title
    </h5>
    <div />
    <svg
      aria-hidden="true"
      aria-label="arrow-down"
      class="c3"
      color="#A141EB"
      fill="currentColor"
      focusable="false"
      height="28"
      style="cursor: pointer;"
      viewBox="0 0 24 24"
      width="28"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        data-name="Layer 2"
      >
        <path
          d="M12 16a1 1 0 01-.64-.23l-6-5a1 1 0 111.28-1.54L12 13.71l5.36-4.32a1 1 0 011.41.15 1 1 0 01-.14 1.46l-6 4.83A1 1 0 0112 16z"
          data-name="arrow-ios-downward"
        />
      </g>
    </svg>
  </div>
  <p
    aria-hidden="true"
    class="c4 c5"
  >
    my children
  </p>
</div>
`;
